File: apps/downloadx/src/hls/session.ts
import path from 'node:path';
import type { DlxContext } from '../types.js';
import { parsePlaylist } from './parser.js';
import type { HlsMediaPlaylist, HlsStream } from './types.js';

/** Resolved media playlist — segments ready to be planned as chunks. */
export interface HlsMediaResolution {
  type: 'media';
  playlist: HlsMediaPlaylist;
}

/** Returned when a master playlist has multiple streams — caller should present
 *  the list to the user; no segments are downloaded in this case. */
export interface HlsMultiStreamResult {
  type: 'multi-stream';
  streams: HlsStream[];
}

export type HlsResolution = HlsMediaResolution | HlsMultiStreamResult;

/**
 * HLS playlist resolver + segment concatenator. Downloading is no longer the
 * session's job — each segment is downloaded as an `isSegment` Chunk by the
 * owning Download. The session only:
 *   - resolves a playlist URL into a media playlist or a multi-stream list,
 *   - registers child downloads for multi-stream master playlists,
 *   - concatenates downloaded segment files into the final output,
 *   - cleans up segment files.
 */
export class HlsSession {
  constructor(
    private readonly id: string,
    private readonly context: DlxContext,
  ) {}

  // ---- playlist resolution -------------------------------------------------

  /** Fetch and parse the playlist at `url`.
   *  - media playlist → { type: 'media', playlist }
   *  - master with >1 stream → { type: 'multi-stream', streams }
   *  - master with 1 stream → resolves that stream's media playlist
   *  Throws on live streams (no #EXT-X-ENDLIST). */
  async resolve(url: string): Promise<HlsResolution> {
    const text = await this.fetchText(url);
    const result = parsePlaylist(text, url);

    if (result.type === 'media') {
      if (result.playlist.isLive) {
        throw new Error('Live HLS streams are not supported');
      }
      return { type: 'media', playlist: result.playlist };
    }

    const streams = result.playlist.streams;
    if (streams.length === 0) throw new Error('HLS master playlist has no streams');

    if (streams.length > 1) {
      return { type: 'multi-stream', streams };
    }

    // Single stream — resolve directly.
    const mediaText = await this.fetchText(streams[0]!.uri);
    const mediaResult = parsePlaylist(mediaText, streams[0]!.uri);
    if (mediaResult.type !== 'media') {
      throw new Error('Expected media playlist, got another master playlist');
    }
    if (mediaResult.playlist.isLive) {
      throw new Error('Live HLS streams are not supported');
    }
    return { type: 'media', playlist: mediaResult.playlist };
  }

  /** Register each stream of a multi-stream master as a separate idle download. */
  async registerStreams(
    streams: HlsStream[],
    baseFilename: string,
    outputPath: string,
  ): Promise<void> {
    const ext = path.extname(baseFilename) || '.ts';
    const stem = path.basename(baseFilename, ext);
    const dir = path.dirname(outputPath);

    for (let i = 0; i < streams.length; i++) {
      const stream = streams[i]!;
      const qualifier =
        stream.resolution ??
        (stream.bandwidth ? `${Math.round(stream.bandwidth / 1000)}kbps` : null) ??
        `stream-${i + 1}`;
      const filename = `${stem} ${qualifier}${ext}`;
      await this.context.addUrl(stream.uri, {
        filename,
        targetPath: dir,
        autoStart: false,
      });
    }
  }

  /** Directory where this download's segment files live. */
  segDir(): string {
    return this.context.io.joinPath(this.context.cachePath, `${this.id}-hls`);
  }

  /** Local file path for segment `index`. */
  segPath(index: number): string {
    return this.context.io.joinPath(this.segDir(), `seg-${String(index).padStart(6, '0')}.ts`);
  }

  // ---- concat --------------------------------------------------------------

  /** Concatenate ordered segment files into `output`. Uses io.concatSegments
   *  (e.g. ffmpeg) when available, otherwise a binary append fallback. */
  async concat(segments: string[], output: string): Promise<void> {
    const io = this.context.io;
    if (io.concatSegments) {
      await io.concatSegments(segments, output);
      return;
    }
    // Binary concat fallback: read each segment and append to output.
    const parts: Uint8Array[] = [];
    for (const seg of segments) {
      parts.push(await io.readFile(seg));
    }
    const total = parts.reduce((acc, p) => acc + p.byteLength, 0);
    const merged = new Uint8Array(total);
    let offset = 0;
    for (const part of parts) {
      merged.set(part, offset);
      offset += part.byteLength;
    }
    await io.writeFile(output, merged);
  }

  // ---- helpers -------------------------------------------------------------

  private async fetchText(url: string): Promise<string> {
    const res = await this.context.io.fetch(url, {
      headers: { Accept: 'application/vnd.apple.mpegurl, application/x-mpegurl, */*' },
    });
    if (!res.ok) throw new Error(`Failed to fetch playlist: HTTP ${res.status} ${url}`);
    return res.text();
  }

  /** Delete all segment files on cancel/error cleanup. */
  async cleanup(segDir: string): Promise<void> {
    try {
      const io = this.context.io;
      // Best-effort: unlink known paths, ignore errors.
      for (let i = 0; ; i++) {
        const p = io.joinPath(segDir, `seg-${String(i).padStart(6, '0')}.ts`);
        if (!(await io.exists(p))) break;
        await io.unlink(p).catch(() => undefined);
      }
    } catch {
      // ignore
    }
  }
}

File: apps/downloadx_dart/lib/src/hls/session.dart
import 'dart:async';

import '../config.dart';
import '../io.dart';
import '../types.dart';
import 'parser.dart';
import 'types.dart';

/// Resolved media playlist — segments ready to be planned as chunks.
class HlsMediaResolution {
  /// The parsed media playlist.
  final HlsMediaPlaylist playlist;

  /// Creates an [HlsMediaResolution].
  HlsMediaResolution(this.playlist);
}

/// Returned when a master playlist has multiple streams — the caller should
/// register each as a child download; no segments are downloaded in this case.
class HlsMultiStreamResult {
  /// The available variant streams.
  final List<HlsStream> streams;

  /// Creates an [HlsMultiStreamResult].
  HlsMultiStreamResult({required this.streams});
}

/// HLS playlist resolver + segment concatenator. Downloading is no longer the
/// session's job — each segment is downloaded as an `isSegment` Chunk by the
/// owning Download. The session only resolves playlists, registers child
/// downloads for multi-stream masters, concatenates segment files, and cleans
/// them up.
class HlsSession {
  /// Download identifier (used for temp segment directory naming).
  final String id;

  /// Manager context providing I/O, config, and the ability to register streams.
  final DlxContext context;

  /// Creates an [HlsSession].
  HlsSession({required this.id, required this.context});

  // ---- playlist resolution -------------------------------------------------

  /// Fetch and parse the playlist at [url].
  /// - media playlist → [HlsMediaResolution]
  /// - master with >1 stream → [HlsMultiStreamResult]
  /// - master with 1 stream → resolves that stream's media playlist
  /// Throws on live streams (no #EXT-X-ENDLIST).
  Future<Object> resolve(String url) async {
    final text = await _fetchText(url);
    final result = parsePlaylist(text, url);

    if (result is HlsMediaResult) {
      if (result.playlist.isLive) {
        throw Exception('Live HLS streams are not supported');
      }
      return HlsMediaResolution(result.playlist);
    }

    final streams = (result as HlsMasterResult).playlist.streams;
    if (streams.isEmpty) throw Exception('HLS master playlist has no streams');

    if (streams.length > 1) return HlsMultiStreamResult(streams: streams);

    // Single stream — resolve directly.
    final mediaText = await _fetchText(streams[0].uri);
    final mediaResult = parsePlaylist(mediaText, streams[0].uri);
    if (mediaResult is! HlsMediaResult) {
      throw Exception('Expected media playlist, got another master playlist');
    }
    if (mediaResult.playlist.isLive) {
      throw Exception('Live HLS streams are not supported');
    }
    return HlsMediaResolution(mediaResult.playlist);
  }

  /// Register each stream of a multi-stream master as a separate idle download.
  Future<void> registerStreams(
    List<HlsStream> streams,
    String baseFilename,
    String outputPath,
  ) async {
    final dotIndex = baseFilename.lastIndexOf('.');
    final ext = dotIndex >= 0 ? baseFilename.substring(dotIndex) : '.ts';
    final stem = dotIndex >= 0 ? baseFilename.substring(0, dotIndex) : baseFilename;
    final dir = _dirname(outputPath);

    for (var i = 0; i < streams.length; i++) {
      final stream = streams[i];
      final String qualifier;
      if (stream.resolution != null) {
        qualifier = stream.resolution!;
      } else if (stream.bandwidth > 0) {
        qualifier = '${(stream.bandwidth / 1000).round()}kbps';
      } else {
        qualifier = 'stream-${i + 1}';
      }
      final filename = '$stem $qualifier$ext';
      await context.addUrl(
        stream.uri,
        DownloadOptions(filename: filename, targetPath: dir, autoStart: false),
      );
    }
  }

  /// Directory where this download's segment files live.
  String segDir() => context.io.joinPath([context.cachePath, '$id-hls']);

  /// Local file path for segment [index].
  String segPath(int index) =>
      context.io.joinPath([segDir(), 'seg-${index.toString().padLeft(6, '0')}.ts']);

  // ---- concat --------------------------------------------------------------

  /// Concatenate ordered segment files into [output]. Uses io.concatSegments
  /// (e.g. ffmpeg) when available, otherwise a binary append fallback.
  Future<void> concat(List<String> segments, String output) async {
    final io = context.io;
    if (io.concatSegments != null) {
      await io.concatSegments!(segments, output);
      return;
    }
    // Binary concat fallback: read each segment and append to output.
    final parts = <List<int>>[];
    for (final seg in segments) {
      parts.add(await io.readFile(seg));
    }
    final merged = parts.fold<List<int>>([], (acc, p) => acc..addAll(p));
    await io.writeFile(output, merged);
  }

  // ---- helpers -------------------------------------------------------------

  Future<String> _fetchText(String url) async {
    final res = await context.io.fetch(
      url,
      FetchInit(headers: {
        'Accept': 'application/vnd.apple.mpegurl, application/x-mpegurl, */*',
      }),
    );
    if (!res.ok) {
      throw Exception('Failed to fetch playlist: HTTP ${res.status} $url');
    }
    return res.text();
  }

  /// Deletes all segment files written to [segDir] by this session.
  Future<void> cleanup(String segDir) async {
    final io = context.io;
    for (var i = 0; ; i++) {
      final p = io.joinPath([segDir, 'seg-${i.toString().padLeft(6, '0')}.ts']);
      if (!await io.exists(p)) break;
      await io.unlink(p).catchError((_) {});
    }
  }

  static String _dirname(String path) {
    final idx = path.lastIndexOf('/');
    if (idx <= 0) return '/';
    return path.substring(0, idx);
  }
}
